pkg/server/middleware/handler_utils.go

```diff
@@ -0,0 +1,42 @@
+// Copyright (C) 2021 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package middleware
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/spacemonkeygo/monkit/v3"
+
+	xhttp "storj.io/minio/cmd/http"
+)
+
+const (
+	// XStorjRequestID is the response request id.
+	XStorjRequestID = "X-Storj-Request-Id"
+	// XStorjParentRequestID is the parent request Id for auth service.
+	XStorjParentRequestID = "X-Storj-Parent-Request-Id"
+)
+
+// AddRequestIds uses XStorjRequestId field to set unique request Ids
+// in the response headers for each request of auth and linksharing service.
+// It also mantains calling service request Id in the XStorjParentRequestId field in auth service response header.
+func AddRequestIds(service string, h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Storing calling service requestId as ParentRequestID for auth service
+		if service == "auth" {
+			if w.Header().Get(XStorjRequestID) != "" {
+				w.Header().Set(XStorjParentRequestID, w.Header().Get(XStorjRequestID))
```

Comment:
if we already have a `X-Storj-Request-Id` in the request headers, shouldn't we pass that straight through instead of putting the existing one in parent? As far as I understand, this middleware should only be generating an ID and setting `X-Storj-Request-Id` if the header is missing, and then setting the one from the request, or the one it just generated to the response header `X-Storj-Request-Id`. I'm not sure what the idea is with the parent header, but maybe I didn't pick up on how this should work.

I would also remove the service check and have this work the same way regardless of who called it.

Reply:
Oh, that's a good idea. I suggested parent request IDs to Aishwarya, so sorry about the confusion. I think Sean's proposal is definitely better.